meja/src/typechecker.ml

```diff
@@ -145,6 +145,101 @@ let check_type ~loc env typ constr_typ =
   | () ->
       ()
 
+(** [is_subtype ~loc env typ ~of_:ctyp] returns whether [typ] is a subtype of
+    [ctyp], instantiating any variables in [ctyp] to those they match with in
+    [typ].
+
+    If this function returns [false], the [ctyp] value *must not* be used,
```

Comment:
a bit scary, would be nice if you could instead return an option or something, but I imagine the mutability here is pretty ingrained

Reply:
I was hitting some performance and consistency problems using a model without mutation. Currently I'm working towards making all the typechecking use mutability, to simplify the algorithms and give us speed and memory-usage improvements.

Once this is done (and the stdlib is working properly), I'm going to add backtracking, so we can recover a working environment when things like this happen. For now, I agree that this is worrying, but I don't have a better solution before the prereqs are in place..

Thanks for approving.